// geckokit/dist/firefox/compatibility.js
const COMPATIBILITY = [
    {
        minimumMajor: 140,
        maximumMajor: 153,
        configure: {
            appBasename: '--with-app-basename',
            branding: '--with-branding',
            artifactBuilds: '--enable-artifact-builds',
            disableWebdriver: '--disable-webdriver',
        },
        make: { objectDirectory: 'MOZ_OBJDIR', autoClobber: 'AUTOCLOBBER' },
        environment: { appName: 'MOZ_APP_NAME' },
        identity: {
            file: 'browser/moz.configure',
            appIdOption: 'MOZ_APP_ID',
            vendorOption: 'MOZ_APP_VENDOR',
        },
        brandingBase: 'browser/branding/unofficial',
        packageManifest: 'browser/installer/package-manifest.in',
        packageEntries: [
            '@RESPATH@/defaults/pref/autoconfig.js',
            '@RESPATH@/geckokit.cfg',
            '@RESPATH@/geckokit/*',
        ],
        bootstrap: { full: 'browser', artifact: 'browser_artifact_mode' },
        workarounds: [],
    },
];
export function firefoxCompatibility(version) {
    const major = Number.parseInt(version, 10);
    const match = COMPATIBILITY.find((record) => major >= record.minimumMajor && major <= record.maximumMajor);
    if (!match) {
        throw new Error(`Firefox ${major} is not supported by this GeckoKit version. Update GeckoKit or select a supported release.`);
    }
    return match;
}
