diff --git a/toolkit/moz.configure b/toolkit/moz.configure
index 593f8b9b9c..7359059d68 100644
--- a/toolkit/moz.configure
+++ b/toolkit/moz.configure
@@ -4481,21 +4481,7 @@ with only_when(~artifact_builds):
     set_define("ONNX_RUNTIME", onnx_runtime_path)
 
 with only_when(artifact_builds):
-
-    @depends(onnxruntime_task_name, build_environment.topsrcdir)
-    @imports(_from="__builtin__", _import="open")
-    @imports("re")
-    def is_onnxruntime_supported_by_taskcluster(onnxruntime_task_name, topsrcdir):
-        task_desc = os.path.join(
-            topsrcdir, "taskcluster", "kinds", "toolchain", "onnx.yml"
-        )
-        with open(task_desc) as fh:
-            content = fh.read()
-            # Those are the runtime we build on taskcluster
-            onnx_tasks = re.findall(r"^(onnxruntime-.*):", content, re.MULTILINE)
-            return onnxruntime_task_name in onnx_tasks
-
-    set_define("ONNX_RUNTIME", is_onnxruntime_supported_by_taskcluster)
+    set_define("ONNX_RUNTIME", False)
 
 set_config(
     "GECKO_TRACE_ENABLE",
