// geckokit/dist/apply.js
import { execFile } from 'node:child_process';
import { mkdir, mkdtemp, readFile, readdir, realpath, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join, posix, resolve } from 'node:path';
import { promisify } from 'node:util';
import fs from 'fs-extra';
import ts from 'typescript';
import { projectPaths } from './project.js';
import { loadBrowserConfig } from './config/load.js';
import { firefoxCompatibility } from './firefox/compatibility.js';
import { readFirefoxLock } from './source.js';
import { featurePolicies } from './features.js';
const { copy } = fs;
const execFileAsync = promisify(execFile);
const INTEGRATION_LINE = 'DIRS += ["geckokit"]';
const PACKAGE_MARKER = '; GeckoKit packaged UI';
const OVERLAY_MANIFEST = '.geckokit-overlay-manifest.json';
const BUILD_MANIFEST = `FINAL_TARGET_FILES.defaults.pref += ["autoconfig.js"]
FINAL_TARGET_FILES.defaults.pref += ["project-prefs.js"]
FINAL_TARGET_FILES += ["geckokit.cfg"]
FINAL_TARGET_FILES.geckokit += ["browser/**"]
FINAL_TARGET_FILES.distribution += ["policies.json"]
`;
const AUTOCONFIG_PREFS = `pref("general.config.filename", "geckokit.cfg");
pref("general.config.obscure_value", 0);
pref("general.config.sandbox_enabled", false);
`;
const AUTOCONFIG = `// GeckoKit managed bootstrap. Edit src/browser, not this file.
const styleSheets = Components.classes["@mozilla.org/content/style-sheet-service;1"]
  .getService(Components.interfaces.nsIStyleSheetService);
const resourceHandler = Services.io.getProtocolHandler("resource")
  .QueryInterface(Components.interfaces.nsIResProtocolHandler);
const resourceDirectory = Services.dirsvc.get("GreD", Components.interfaces.nsIFile);
resourceDirectory.append("geckokit");
resourceHandler.setSubstitution("geckokit", Services.io.newFileURI(resourceDirectory));

for (const name of ["browser.css", "content.css"]) {
  const file = Services.dirsvc.get("GreD", Components.interfaces.nsIFile);
  file.append("geckokit");
  file.append("styles");
  file.append(name);
  if (file.exists()) {
    styleSheets.loadAndRegisterSheet(Services.io.newFileURI(file), styleSheets.USER_SHEET);
  }
}

const startGeckoKit = async window => {
  Services.obs.removeObserver(startGeckoKit, "browser-delayed-startup-finished");
  const verificationEnabled = Services.prefs.getBoolPref("geckokit.verification.enabled", false);
  const finishVerification = passed => {
    if (!verificationEnabled) {
      return;
    }
    Services.prefs.setBoolPref("geckokit.verification.passed", passed);
    Services.prefs.savePrefFile(null);
    Services.startup.quit(Components.interfaces.nsIAppStartup.eAttemptQuit);
  };
  try {
    await ChromeUtils.importESModule("resource://geckokit/startup.sys.mjs").startup(window);
    finishVerification(true);
  } catch (error) {
    Components.utils.reportError(error);
    finishVerification(false);
  }
};
Services.obs.addObserver(startGeckoKit, "browser-delayed-startup-finished");
`;
async function gitCheck(source, args, env = process.env) {
    try {
        await execFileAsync('git', ['-C', source, ...args], { env });
        return true;
    }
    catch {
        return false;
    }
}
async function listOverlayFiles(directory, prefix = '') {
    const files = [];
    for (const entry of await readdir(resolve(directory, prefix), { withFileTypes: true })) {
        const path = posix.join(prefix, entry.name);
        if (entry.isDirectory())
            files.push(...await listOverlayFiles(directory, path));
        else if (entry.isFile() && path !== 'README.md')
            files.push(path);
    }
    return files.sort();
}
async function readOverlayManifest(path) {
    try {
        const value = JSON.parse(await readFile(path, 'utf8'));
        if (!Array.isArray(value) ||
            !value.every((item) => typeof item === 'string' &&
                item.length > 0 &&
                item !== '.' &&
                item !== '..' &&
                posix.normalize(item) === item &&
                !posix.isAbsolute(item) &&
                !item.startsWith('../') &&
                !item.includes('\\'))) {
            throw new Error('expected safe relative path array');
        }
        return value;
    }
    catch (error) {
        if (error.code === 'ENOENT')
            return [];
        throw new Error(`Invalid ${OVERLAY_MANIFEST}: ${error instanceof Error ? error.message : String(error)}`);
    }
}
async function patchFiles(source, patches) {
    const files = new Set();
    for (const patch of patches) {
        const { stdout } = await execFileAsync('git', ['-C', source, 'apply', '--numstat', '-z', patch], { encoding: 'utf8' });
        const records = stdout.split('\0');
        for (let index = 0; index < records.length - 1; index++) {
            const path = records[index].split('\t')[2];
            if (path)
                files.add(path);
            else {
                files.add(records[++index]);
                files.add(records[++index]);
            }
        }
    }
    return [...files];
}
function setImpliedIdentity(source, option, value) {
    const pattern = new RegExp(`^imply_option\\("${option}", .+\\)$`, 'gm');
    const matches = source.match(pattern);
    if (matches?.length !== 1) {
        throw new Error(`Firefox identity option ${option} changed; update GeckoKit compatibility data.`);
    }
    return source.replace(pattern, `imply_option("${option}", ${JSON.stringify(value)})`);
}
async function patchPlan(source, patchDirectory) {
    const patches = (await readdir(patchDirectory, { withFileTypes: true }))
        .filter((entry) => entry.isFile() && entry.name.endsWith('.patch'))
        .map((entry) => resolve(patchDirectory, entry.name))
        .sort();
    if (patches.length === 0)
        return [];
    if (await gitCheck(source, ['apply', '--reverse', '--check', ...[...patches].reverse()]))
        return [];
    const pending = [];
    const indexDirectory = await mkdtemp(join(tmpdir(), 'geckokit-index-'));
    const indexEnvironment = { ...process.env, GIT_INDEX_FILE: resolve(indexDirectory, 'index') };
    try {
        await execFileAsync('git', ['-C', source, 'read-tree', 'HEAD'], { env: indexEnvironment });
        const files = await patchFiles(source, patches);
        const existingOrTracked = [];
        for (const path of files) {
            if (await fs.pathExists(resolve(source, path)) ||
                await gitCheck(source, ['cat-file', '-e', `HEAD:${path}`]))
                existingOrTracked.push(path);
        }
        if (existingOrTracked.length > 0) {
            await execFileAsync('git', ['-C', source, 'add', '-A', '--', ...existingOrTracked], { env: indexEnvironment });
        }
        const applied = new Set();
        let peelingAppliedPrefix = false;
        for (let index = patches.length - 1; index >= 0; index--) {
            const patch = patches[index];
            if (await gitCheck(source, ['apply', '--cached', '--reverse', '--check', patch], indexEnvironment)) {
                await execFileAsync('git', ['-C', source, 'apply', '--cached', '--reverse', patch], { env: indexEnvironment });
                applied.add(index);
                peelingAppliedPrefix = true;
            }
            else if (peelingAppliedPrefix) {
                throw new Error(`Patch stack is not an applied prefix: ${patches[index]}`);
            }
        }
        for (let index = 0; index < patches.length; index++) {
            const patch = patches[index];
            if (!(await gitCheck(source, ['apply', '--cached', '--check', patch], indexEnvironment))) {
                throw new Error(`Patch does not apply: ${patch}`);
            }
            await execFileAsync('git', ['-C', source, 'apply', '--cached', patch], { env: indexEnvironment });
            if (!applied.has(index))
                pending.push(patch);
        }
        return pending;
    }
    finally {
        await rm(indexDirectory, { recursive: true, force: true });
    }
}
export async function validateProjectPatches(cwd = process.cwd()) {
    const paths = await projectPaths(cwd);
    const config = await loadBrowserConfig(cwd);
    await patchPlan(await realpath(paths.firefoxSource), resolve(paths.projectDirectory, config.patches.directory));
}
export async function validateProjectIdentity(cwd = process.cwd()) {
    const paths = await projectPaths(cwd);
    const config = await loadBrowserConfig(cwd);
    const compatibility = firefoxCompatibility((await readFirefoxLock(cwd)).version);
    const identity = await readFile(resolve(paths.firefoxSource, compatibility.identity.file), 'utf8');
    const expected = [
        `imply_option("${compatibility.identity.vendorOption}", ${JSON.stringify(config.browser.vendor)})`,
        `imply_option("${compatibility.identity.appIdOption}", ${JSON.stringify(config.browser.applicationId)})`,
    ];
    if (!expected.every((line) => identity.includes(line))) {
        throw new Error('Firefox source identity does not match browser.config.json');
    }
}
export async function applyProject(cwd = process.cwd()) {
    const paths = await projectPaths(cwd);
    const config = await loadBrowserConfig(cwd);
    const source = await realpath(paths.firefoxSource);
    const compatibility = firefoxCompatibility((await readFirefoxLock(cwd)).version);
    const overlayDirectory = resolve(paths.projectDirectory, config.overlays.directory);
    const overlayManifestPath = resolve(source, OVERLAY_MANIFEST);
    const currentOverlays = await listOverlayFiles(overlayDirectory);
    const previousOverlays = await readOverlayManifest(overlayManifestPath);
    for (const path of previousOverlays.filter((path) => !currentOverlays.includes(path))) {
        if (await gitCheck(source, ['cat-file', '-e', `HEAD:${path}`])) {
            await execFileAsync('git', ['-C', source, 'checkout', 'HEAD', '--', path]);
        }
        else {
            await rm(resolve(source, path), { force: true });
        }
    }
    await copy(overlayDirectory, source, {
        overwrite: true,
        filter: (path) => path !== resolve(overlayDirectory, 'README.md'),
    });
    await writeFile(overlayManifestPath, `${JSON.stringify(currentOverlays, null, 2)}\n`);
    if (config.branding.enabled) {
        const target = resolve(source, `browser/branding/geckokit-${config.browser.binaryName}`);
        await copy(resolve(source, compatibility.brandingBase), target, { overwrite: true });
        await copy(resolve(paths.projectDirectory, config.branding.directory), target, {
            overwrite: true,
        });
    }
    const identityPath = resolve(source, compatibility.identity.file);
    let identity = await readFile(identityPath, 'utf8');
    identity = setImpliedIdentity(identity, compatibility.identity.vendorOption, config.browser.vendor);
    identity = setImpliedIdentity(identity, compatibility.identity.appIdOption, config.browser.applicationId);
    await writeFile(identityPath, identity);
    const patchDirectory = resolve(paths.projectDirectory, config.patches.directory);
    const pending = await patchPlan(source, patchDirectory);
    for (const patch of pending) {
        await execFileAsync('git', ['-C', source, 'apply', patch]);
    }
    const managed = resolve(source, 'browser/app/geckokit');
    await mkdir(managed, { recursive: true });
    const prefsDirectory = resolve(paths.projectDirectory, 'prefs');
    const projectPrefs = await Promise.all((await readdir(prefsDirectory))
        .filter((name) => name.endsWith('.js'))
        .sort()
        .map((name) => readFile(resolve(prefsDirectory, name), 'utf8')));
    await rm(resolve(managed, 'browser'), { recursive: true, force: true });
    const browserSource = resolve(paths.projectDirectory, 'src/browser');
    const browserTarget = resolve(managed, 'browser');
    await Promise.all([
        writeFile(resolve(managed, 'moz.build'), BUILD_MANIFEST),
        writeFile(resolve(managed, 'autoconfig.js'), AUTOCONFIG_PREFS),
        writeFile(resolve(managed, 'geckokit.cfg'), AUTOCONFIG),
        writeFile(resolve(managed, 'project-prefs.js'), `${projectPrefs.join('\n').trimEnd()}\n`),
        writeFile(resolve(managed, 'policies.json'), `${JSON.stringify({ policies: featurePolicies(config.features.selected) }, null, 2)}\n`),
        copy(browserSource, browserTarget, {
            overwrite: true,
            filter: (path) => !/\.m?ts$/.test(path),
        }),
    ]);
    for (const path of await listOverlayFiles(browserSource)) {
        if (path.endsWith('.d.ts') || (!path.endsWith('.mts') && !path.endsWith('.ts')))
            continue;
        const output = ts.transpileModule(await readFile(resolve(browserSource, path), 'utf8'), {
            compilerOptions: { module: ts.ModuleKind.ESNext, target: ts.ScriptTarget.ES2022 },
            fileName: path,
            reportDiagnostics: true,
        });
        const error = output.diagnostics?.find((diagnostic) => diagnostic.category === ts.DiagnosticCategory.Error);
        if (error) {
            throw new Error(`TypeScript error in ${path}: ${ts.flattenDiagnosticMessageText(error.messageText, '\n')}`);
        }
        const destination = path.endsWith('.mts') ? `${path.slice(0, -4)}.mjs` : `${path.slice(0, -3)}.js`;
        await mkdir(resolve(browserTarget, posix.dirname(destination)), { recursive: true });
        await writeFile(resolve(browserTarget, destination), output.outputText);
    }
    const appBuildPath = resolve(source, 'browser/app/moz.build');
    const appBuild = await readFile(appBuildPath, 'utf8');
    if (!new RegExp(`^${INTEGRATION_LINE.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}$`, 'm').test(appBuild)) {
        await writeFile(appBuildPath, `${appBuild.trimEnd()}\n\n# GeckoKit packaged UI\n${INTEGRATION_LINE}\n`);
    }
    const packageManifestPath = resolve(source, compatibility.packageManifest);
    const packageManifest = await readFile(packageManifestPath, 'utf8');
    if (!packageManifest.includes(PACKAGE_MARKER)) {
        await writeFile(packageManifestPath, `${packageManifest.trimEnd()}\n\n${PACKAGE_MARKER}\n${compatibility.packageEntries.join('\n')}\n`);
    }
}
