// geckokit/dist/cli.js
#!/usr/bin/env node
import { confirm, isCancel, multiselect, select, spinner, text } from '@clack/prompts';
import { Command } from 'commander';
import fs from 'fs-extra';
import { basename } from 'node:path';
import { applyProject } from './apply.js';
import { BrowserConfigSchema } from './config/schema.js';
import { createProject } from './create.js';
import { doctorProject } from './doctor.js';
import { runMach } from './mach.js';
import { prepareProject } from './prepare.js';
import { projectPaths } from './project.js';
import { setupProject } from './setup.js';
import { downloadFirefox } from './source.js';
import { runWorkflow } from './workflow.js';
import { verifyProject } from './verify.js';
import { VERSION } from './version.js';
import { exportFirefoxPatch } from './patches.js';
import { FEATURE_CATALOG, FEATURE_PRESETS, } from './features.js';
const { pathExists } = fs;
function answer(value) {
    if (isCancel(value))
        throw new Error('Cancelled.');
    return value;
}
async function loading(message, done, task) {
    const indicator = spinner();
    indicator.start(message);
    try {
        const result = await task();
        indicator.stop(done);
        return result;
    }
    catch (error) {
        indicator.error(`${message} failed.`);
        throw error;
    }
}
async function createConfig(projectName, flags) {
    const interactive = Boolean(process.stdin.isTTY && process.stdout.isTTY);
    const askText = async (message, initialValue) => answer(await text({ message, initialValue }));
    const target = projectName ?? (interactive
        ? await askText('Project directory', 'my-browser')
        : (() => { throw new Error('Missing project name. Use `geckokit create <name>`.'); })());
    const name = target === '.' ? basename(process.cwd()) : target;
    const defaultName = name[0].toUpperCase() + name.slice(1);
    const browserName = flags.browserName ?? (interactive ? await askText('Browser name', defaultName) : defaultName);
    const binaryName = flags.binaryName ?? (interactive ? await askText('Binary name', name) : name);
    const applicationId = flags.applicationId ?? (interactive ? await askText('Application ID', `dev.geckokit.${name}`) : `dev.geckokit.${name}`);
    const vendor = flags.vendor ?? (interactive ? await askText('Vendor', browserName) : browserName);
    const channel = flags.channel ?? (interactive ? answer(await select({
        message: 'Firefox channel',
        options: [{ value: 'release', label: 'Stable release' }, { value: 'esr', label: 'ESR' }],
    })) : 'release');
    const featurePreset = flags.featureProfile ?? (interactive ? answer(await select({
        message: 'Feature profile',
        options: [
            { value: 'default', label: 'Firefox default', hint: 'No features disabled' },
            { value: 'privacy', label: 'Privacy focused', hint: 'Official policies; no claimed size reduction' },
            { value: 'minimal', label: 'Minimal browser', hint: 'Privacy policies plus verified build removal' },
            { value: 'custom', label: 'Custom', hint: 'Choose tested features' },
        ],
    })) : 'default');
    const selectedFeatures = featurePreset === 'custom' && interactive
        ? answer(await multiselect({
            message: 'Features to disable or remove',
            options: FEATURE_CATALOG.map((feature) => ({
                value: feature.id,
                label: feature.label,
                hint: feature.hint,
            })),
            required: false,
        }))
        : [...(FEATURE_PRESETS[featurePreset] ?? [])];
    const needsFullBuild = selectedFeatures.includes('webdriver');
    const mode = flags.mode ?? (interactive ? answer(await select({
        message: 'Build mode',
        options: needsFullBuild
            ? [
                { value: 'full', label: 'Full build', hint: 'Required by selected feature removals' },
                { value: 'artifact', label: 'Artifact build', hint: 'Fast UI work; incompatible with selected removals' },
            ]
            : [
                { value: 'artifact', label: 'Artifact build', hint: 'Recommended for UI, CSS, JS, and branding' },
                { value: 'full', label: 'Full build', hint: 'Required for Gecko, native code, and release packages' },
            ],
    })) : (needsFullBuild ? 'full' : 'artifact'));
    const branding = flags.branding ?? (interactive ? answer(await confirm({ message: 'Enable custom branding?', initialValue: false })) : false);
    const language = flags.language ?? (interactive ? answer(await select({
        message: 'Browser code language',
        options: [
            { value: 'javascript', label: 'JavaScript', hint: 'Native Firefox modules' },
            { value: 'typescript', label: 'TypeScript', hint: 'Transpiled to Firefox modules during apply' },
        ],
    })) : 'javascript');
    return {
        projectName: target,
        config: BrowserConfigSchema.parse({
            browser: { name: browserName, binaryName, vendor, applicationId },
            firefox: { channel, version: 'latest' },
            branding: { enabled: branding, directory: 'branding' },
            patches: { directory: 'patches' },
            overlays: { directory: 'src/firefox' },
            features: { preset: featurePreset, selected: selectedFeatures },
            build: { mode },
            language,
        }),
    };
}
function workflowCommand(program, name, description) {
    program.command(`${name} [args...]`)
        .description(description)
        .allowUnknownOption(true)
        .action(async (args = []) => { process.exitCode = await runWorkflow(name, args); });
}
const program = new Command()
    .name('geckokit')
    .description('Build custom Firefox-based browsers with Mozilla mach')
    .version(VERSION)
    .option('--verbose', 'show stack traces');
program.command('create [name]')
    .description('scaffold a custom browser project')
    .option('--browser-name <name>')
    .option('--binary-name <name>')
    .option('--application-id <id>')
    .option('--vendor <name>')
    .option('--channel <channel>', 'release or esr')
    .option('--mode <mode>', 'full or artifact')
    .option('--feature-profile <profile>', 'default, privacy, minimal, or custom')
    .option('--language <language>', 'javascript or typescript')
    .option('--branding', 'enable custom branding')
    .action(async (name, flags) => {
    const selected = await createConfig(name, flags);
    const directory = await loading('Creating project', 'Project created.', () => createProject(selected.projectName, process.cwd(), selected.config));
    const prefix = selected.projectName === '.' ? '' : `cd ${selected.projectName} && `;
    const next = VERSION === '0.0.0'
        ? `${prefix}geckokit setup`
        : `${prefix}npm install && npx geckokit setup`;
    console.log(`Created ${directory}\nNext: ${next}`);
});
program.command('setup')
    .description('download, bootstrap, configure, and apply Firefox source')
    .action(async () => {
    const lock = await setupProject();
    console.log(`Firefox ${lock.version} ready.`);
});
program.command('doctor')
    .description('diagnose the local build environment')
    .action(async () => {
    const checks = await loading('Checking environment', 'Environment checked.', () => doctorProject());
    for (const check of checks) {
        console.log(`${check.ok ? 'OK' : 'FAIL'} ${check.name}: ${check.detail}${check.fix ? `\n  Fix: ${check.fix}` : ''}`);
    }
    if (checks.some((check) => !check.ok))
        process.exitCode = 1;
});
program.command('verify')
    .description('launch the built browser and verify project startup')
    .action(async () => {
    await verifyProject();
    console.log('Firefox startup verified.');
});
program.command('sync')
    .description('restore missing locked source or reapply project customizations')
    .action(async () => {
    const paths = await projectPaths();
    if (await pathExists(paths.firefoxSource))
        await prepareProject();
    else
        await setupProject();
});
const applyWithProgress = () => loading('Applying project files and patches', 'Project files and patches applied.', () => applyProject());
program.command('apply').description('apply overlays, branding, and patches').action(applyWithProgress);
program.command('export <path>').description('save a Firefox source edit as a project patch').action(async (path) => {
    const target = await loading('Exporting patch', 'Patch exported.', () => exportFirefoxPatch(path));
    console.log(`Exported ${target}`);
});
program.command('import').description('apply project files and patches to Firefox source').action(applyWithProgress);
program.command('clean').description('remove Mozilla build outputs').action(async () => {
    process.exitCode = await runMach(['clobber']);
});
program.command('download').description('download locked or latest stable Firefox source').action(async () => {
    const lock = await loading('Downloading Firefox', 'Firefox downloaded.', () => downloadFirefox());
    console.log(`Downloaded Firefox ${lock.version} (${lock.revision.slice(0, 12)}).`);
});
program.command('bootstrap [args...]').allowUnknownOption(true).action(async (args = []) => {
    process.exitCode = await runMach(['bootstrap', ...args]);
});
workflowCommand(program, 'build', 'apply and build with mach');
workflowCommand(program, 'run', 'apply and launch the development browser');
workflowCommand(program, 'package', 'build and create Mozilla release packages');
program.command('mach <command> [args...]').allowUnknownOption(true).action(async (command, args = []) => {
    process.exitCode = await runMach([command, ...args]);
});
program.parseAsync().catch((error) => {
    const value = error instanceof Error ? error : new Error(String(error));
    console.error(program.opts().verbose ? value.stack : value.message);
    process.exitCode = 1;
});
