// geckokit/dist/features.js
export const FEATURE_IDS = [
    'pocket',
    'studies',
    'telemetry-upload',
    'firefox-accounts',
    'screenshots',
    'webdriver',
];
export const FEATURE_CATALOG = [
    { id: 'pocket', label: 'Disable Pocket', hint: 'Privacy/UI; no size reduction', policy: 'DisablePocket' },
    { id: 'studies', label: 'Disable Firefox Studies', hint: 'Privacy; no size reduction', policy: 'DisableFirefoxStudies' },
    { id: 'telemetry-upload', label: 'Disable telemetry upload', hint: 'Privacy; no size reduction', policy: 'DisableTelemetry' },
    { id: 'firefox-accounts', label: 'Disable Firefox Accounts', hint: 'Privacy/UI; disables Sync', policy: 'DisableFirefoxAccounts' },
    { id: 'screenshots', label: 'Disable Firefox Screenshots', hint: 'UI; no size reduction', policy: 'DisableFirefoxScreenshots' },
    { id: 'webdriver', label: 'Remove WebDriver components', hint: 'Smaller build; disables browser automation', requiresFullBuild: true },
];
export const FEATURE_PRESETS = {
    default: [],
    privacy: ['pocket', 'studies', 'telemetry-upload'],
    minimal: ['pocket', 'studies', 'telemetry-upload', 'firefox-accounts', 'screenshots', 'webdriver'],
    custom: [],
};
export function featurePolicies(selected) {
    return Object.fromEntries(FEATURE_CATALOG
        .filter((feature) => selected.includes(feature.id) && feature.policy)
        .map((feature) => [feature.policy, true]));
}
