// geckokit/dist/mach.js
import { constants } from 'node:os';
import { access, copyFile, mkdir, readFile, realpath, writeFile } from 'node:fs/promises';
import { constants as fsConstants } from 'node:fs';
import { resolve, win32 } from 'node:path';
import { spawn } from 'node:child_process';
import { projectPaths } from './project.js';
async function installProfile(projectDirectory) {
    const profile = resolve(projectDirectory, '.geckokit-profile');
    const chrome = resolve(profile, 'chrome');
    const styles = resolve(projectDirectory, 'src/browser/styles');
    await mkdir(chrome, { recursive: true });
    await Promise.all([
        copyFile(resolve(styles, 'browser.css'), resolve(chrome, 'userChrome.css')),
        copyFile(resolve(styles, 'content.css'), resolve(chrome, 'userContent.css')),
        readFile(resolve(projectDirectory, 'prefs/browser.js'), 'utf8').then((prefs) => writeFile(resolve(profile, 'user.js'), prefs.replace(/^pref\(/gm, 'user_pref('))),
    ]);
    return profile;
}
function hasProfileArgument(args) {
    return args.some((arg) => arg === '--profile' ||
        arg.startsWith('--profile=') ||
        arg === '-profile' ||
        arg === '-P');
}
export function machInvocation(firefoxSource, args, platform = process.platform) {
    if (platform === 'win32') {
        return {
            command: 'powershell.exe',
            args: [
                '-NoProfile',
                '-ExecutionPolicy',
                'Bypass',
                '-File',
                win32.join(firefoxSource, 'mach.ps1'),
                ...args,
            ],
        };
    }
    return { command: resolve(firefoxSource, 'mach'), args };
}
export async function runMach(args, cwd = process.cwd(), signal) {
    if (args.length === 0) {
        throw new Error('Missing mach command');
    }
    const paths = await projectPaths(cwd);
    const projectDirectory = paths.projectDirectory;
    const firefoxSource = await realpath(paths.firefoxSource);
    const mach = resolve(firefoxSource, process.platform === 'win32' ? 'mach.ps1' : 'mach');
    try {
        await access(mach, process.platform === 'win32' ? fsConstants.F_OK : fsConstants.X_OK);
    }
    catch {
        throw new Error(`Mozilla mach executable not found: ${mach}`);
    }
    const commandArgs = [...args];
    if (commandArgs[0] === 'run' && !hasProfileArgument(commandArgs)) {
        commandArgs.push('--profile', await installProfile(projectDirectory));
    }
    const invocation = machInvocation(firefoxSource, commandArgs);
    if (signal?.aborted)
        return 130;
    return new Promise((resolveExit, reject) => {
        const child = spawn(invocation.command, invocation.args, {
            cwd: firefoxSource,
            env: { ...process.env, MOZCONFIG: resolve(projectDirectory, 'mozconfig') },
            stdio: 'inherit',
        });
        const forwardInterrupt = () => child.kill('SIGINT');
        const forwardTermination = () => child.kill('SIGTERM');
        const abort = () => child.kill('SIGTERM');
        if (!signal) {
            process.on('SIGINT', forwardInterrupt);
            process.on('SIGTERM', forwardTermination);
        }
        signal?.addEventListener('abort', abort, { once: true });
        const cleanup = () => {
            if (!signal) {
                process.off('SIGINT', forwardInterrupt);
                process.off('SIGTERM', forwardTermination);
            }
            signal?.removeEventListener('abort', abort);
        };
        child.once('error', (error) => {
            cleanup();
            reject(error);
        });
        child.once('close', (code, signal) => {
            cleanup();
            if (code !== null) {
                resolveExit(code);
                return;
            }
            resolveExit(128 + (signal ? constants.signals[signal] : 0));
        });
    });
}
