// geckokit/dist/source.js
import { spawn, execFile } from 'node:child_process';
import { mkdir, readFile, rename, rm, stat, writeFile } from 'node:fs/promises';
import { randomUUID } from 'node:crypto';
import { dirname, resolve } from 'node:path';
import { promisify } from 'node:util';
import { projectPaths } from './project.js';
const FIREFOX_VERSIONS_URL = 'https://product-details.mozilla.org/1.0/firefox_versions.json';
const FIREFOX_REPOSITORY = 'https://github.com/mozilla-firefox/firefox.git';
const execFileAsync = promisify(execFile);
async function readLock(path) {
    try {
        const value = JSON.parse(await readFile(path, 'utf8'));
        if (typeof value.version !== 'string' ||
            value.tag !== releaseTag(value.version) ||
            typeof value.revision !== 'string' ||
            !/^[0-9a-f]{40,64}$/.test(value.revision)) {
            throw new Error('invalid version, tag, or revision');
        }
        return value;
    }
    catch (error) {
        if (error.code === 'ENOENT')
            return undefined;
        throw new Error(`Invalid geckokit.lock.json: ${error instanceof Error ? error.message : String(error)}`);
    }
}
export async function readFirefoxLock(cwd = process.cwd()) {
    const { projectDirectory } = await projectPaths(cwd);
    const lock = await readLock(resolve(projectDirectory, 'geckokit.lock.json'));
    if (!lock)
        throw new Error('Firefox is not locked. Run `geckokit setup` first.');
    return lock;
}
export async function latestFirefoxVersion(fetcher = fetch, channel = 'release') {
    const response = await fetcher(FIREFOX_VERSIONS_URL);
    if (!response.ok) {
        throw new Error(`Mozilla version request failed: HTTP ${response.status}`);
    }
    const value = (await response.json());
    const key = channel === 'esr' ? 'FIREFOX_ESR' : 'LATEST_FIREFOX_VERSION';
    if (typeof value[key] !== 'string') {
        throw new Error(`Mozilla response missing ${key}`);
    }
    releaseTag(value[key]);
    return value[key];
}
export function releaseTag(version) {
    if (!/^\d+(?:\.\d+)+(?:esr)?$/.test(version)) {
        throw new Error(`Invalid Firefox version: ${version}`);
    }
    return `FIREFOX_${version.replaceAll('.', '_')}_RELEASE`;
}
function run(command, args) {
    return new Promise((resolveRun, reject) => {
        const child = spawn(command, args, { stdio: 'inherit' });
        child.once('error', reject);
        child.once('close', (code) => {
            if (code === 0)
                resolveRun();
            else
                reject(new Error(`${command} exited with code ${code}`));
        });
    });
}
export async function downloadFirefox(cwd = process.cwd(), version, repository = FIREFOX_REPOSITORY) {
    const { projectDirectory, firefoxSource: target } = await projectPaths(cwd);
    try {
        await stat(target);
        throw new Error(`Firefox source already exists: ${target}`);
    }
    catch (error) {
        if (error.code !== 'ENOENT')
            throw error;
    }
    const lockPath = resolve(projectDirectory, 'geckokit.lock.json');
    const storedLock = await readLock(lockPath);
    if (storedLock && version !== undefined && version !== storedLock.version) {
        throw new Error(`Firefox ${storedLock.version} is locked; remove lockfile to select ${version}`);
    }
    const resolvedVersion = storedLock?.version ?? version ?? (await latestFirefoxVersion());
    const tag = storedLock?.tag ?? releaseTag(resolvedVersion);
    await mkdir(dirname(target), { recursive: true });
    const temporaryTarget = `${target}.download-${randomUUID()}`;
    let installed = false;
    try {
        await run('git', ['clone', '--depth', '1', '--branch', tag, repository, temporaryTarget]);
        const { stdout } = await execFileAsync('git', ['-C', temporaryTarget, 'rev-parse', 'HEAD']);
        const revision = stdout.trim();
        if (storedLock && revision !== storedLock.revision) {
            throw new Error(`Firefox tag ${tag} no longer matches locked revision ${storedLock.revision}`);
        }
        const lock = storedLock ?? { version: resolvedVersion, tag, revision };
        await rename(temporaryTarget, target);
        installed = true;
        if (!storedLock) {
            await writeFile(lockPath, `${JSON.stringify(lock, null, 2)}\n`, { flag: 'wx' });
        }
        return lock;
    }
    catch (error) {
        await rm(temporaryTarget, { recursive: true, force: true });
        if (installed)
            await rm(target, { recursive: true, force: true });
        throw error;
    }
}
