// geckokit/dist/config/schema.js
import { z } from 'zod';
import { FEATURE_IDS } from '../features.js';
export const BrowserConfigSchema = z.object({
    browser: z.object({
        name: z.string().trim().min(1).regex(/^[^\r\n]+$/),
        binaryName: z.string().regex(/^[a-z0-9][a-z0-9-]*$/),
        vendor: z.string().trim().min(1).regex(/^[^\r\n]+$/),
        applicationId: z.string().regex(/^[a-zA-Z][a-zA-Z0-9.-]+$/),
    }),
    firefox: z.object({
        channel: z.enum(['release', 'esr']),
        version: z.string().regex(/^(?:latest|\d+(?:\.\d+){1,2}(?:esr)?)$/),
    }),
    branding: z.object({
        enabled: z.boolean().default(false),
        directory: z.string().min(1).default('branding'),
    }),
    patches: z.object({ directory: z.string().min(1).default('patches') }),
    overlays: z.object({ directory: z.string().min(1).default('src/firefox') }),
    features: z.object({
        preset: z.enum(['default', 'privacy', 'minimal', 'custom']),
        selected: z.array(z.enum(FEATURE_IDS)).default([]),
    }).default({ preset: 'default', selected: [] }),
    build: z.object({ mode: z.enum(['full', 'artifact']) }),
    language: z.enum(['javascript', 'typescript']).default('javascript'),
}).superRefine((config, context) => {
    if (config.build.mode === 'artifact' && config.features.selected.includes('webdriver')) {
        context.addIssue({
            code: 'custom',
            path: ['features', 'selected'],
            message: 'WebDriver removal requires full build mode.',
        });
    }
});
