// geckokit/dist/scaffold/templates.js
import { customizationGuide } from './customizing.js';
import { VERSION } from '../version.js';
export function scaffoldTemplates(name, selected) {
    const displayName = selected?.browser.name ?? name[0].toUpperCase() + name.slice(1);
    const vendor = selected?.browser.vendor ?? displayName;
    const json = (value) => `${JSON.stringify(value, null, 2)}\n`;
    const shellQuote = (value) => `'${value.replaceAll("'", "'\\''")}'`;
    const language = selected?.language ?? 'javascript';
    const startupFile = language === 'typescript' ? 'startup.sys.mts' : 'startup.sys.mjs';
    return {
        'README.md': `# ${displayName}

Custom Firefox browser built with GeckoKit.

Complete tutorial: \`docs/customizing.md\`

## Start here

\`\`\`bash
npm install
npm run setup
npm run build
npm run verify
npm run dev
\`\`\`

Use artifact mode for JS, CSS, prefs, and branding. Use full mode for Gecko, native code, WebIDL, or release packages.

## Main directories

\`\`\`text
src/browser/                 ${displayName}-specific features and UI
  startup.sys.${language === 'typescript' ? 'mts' : 'mjs'}        small entry point; imports feature modules
  features/                  one module or folder per feature
  views/                     packaged HTML, CSS, and JS
  assets/                    icons and images
  styles/                    browser chrome and content CSS
src/firefox/                 complete files mirrored onto Firefox source paths
prefs/                       native Firefox default preferences
patches/                     small ordered edits to upstream Firefox files
branding/                    product names, icons, and platform assets
.geckokit/firefox/           generated working Firefox checkout and build tree
\`\`\`

This follows Zen Browser's useful architecture: product features stay under \`src/\`, preferences stay separate, and upstream Firefox changes remain visible as mirrored files or patches.

## Choose the smallest customization layer

1. Appearance: edit \`src/browser/styles/\`.
2. Existing Firefox behavior: edit \`prefs/browser.js\`.
3. New UI or behavior: add a module under \`src/browser/features/\`, then import it from \`src/browser/${startupFile}\`.
4. Own a complete upstream file: copy it to the same path under \`src/firefox/\`.
5. Change a few upstream lines: export a patch into \`patches/\`.

After the first build, run \`npm run dev\`. It watches project files, restarts directly for profile CSS, uses \`mach build faster\` for frontend changes, and uses a full build for config, patches, or native/build-system files.

## Patch Firefox source

Use [Searchfox](https://searchfox.org/mozilla-central/source) to find a file. Its path after \`mozilla-central/source/\` matches the path inside \`.geckokit/firefox/\`.

\`\`\`bash
# 1. Edit the generated working copy.
$EDITOR .geckokit/firefox/<path>

# 2. Save that Git diff as a durable project patch.
npm run export -- <path>

# 3. Reapply the ordered patch stack and rebuild.
npm run import
npm run build:fast
\`\`\`

Never leave lasting work only in \`.geckokit/firefox/\`; \`setup\` can recreate it. Keep durable changes in \`src/firefox/\` or \`patches/\`. Use \`npm run build\` instead for C++, Rust, WebIDL, configure, or build-system changes.

## Welcome UI

The first launch shows a native notification bar from \`src/browser/${startupFile}\`.

- Edit the marked onboarding block to change it.
- Remove the block to disable it.
- Reset \`geckokit.onboarding.seen\` in \`about:config\` to show it again.

## Release

Set \`build.mode\` to \`full\` in \`browser.config.json\`, replace all branding assets, rerun \`npm run setup\`, then run \`npm run package\`.
`,
        'AGENTS.md': `# Project instructions

Custom Firefox browser: ${displayName}. Built with GeckoKit.

## Use the smallest customization layer

1. Appearance: \`src/browser/styles/\`.
2. Firefox defaults: \`prefs/browser.js\`.
3. Product features: one module or folder under \`src/browser/features/\`, imported from \`src/browser/${startupFile}\`.
4. Complete upstream file ownership: matching path under \`src/firefox/\`.
5. Small upstream edits: ordered patches under \`patches/\`.

Never keep lasting work only in \`.geckokit/firefox/\`; it is a generated working checkout.

## Commands

- \`npm run setup\`: download, bootstrap, configure, and apply Firefox.
- \`npm run dev\`: watch changes, choose the safe build, and restart Firefox.
- \`npm run build:fast\`: rebuild frontend JS, CSS, prefs, views, overlays, and frontend patches.
- \`npm run build\`: build native code, WebIDL, configure, or build-system changes.
- \`npm start\`: run the browser.
- \`npm run verify\`: launch an isolated profile and verify startup.
- \`npm run export -- <path>\`: export one edited Firefox file as a patch.
- \`npm run import\`: reapply overlays and the ordered patch stack.

## Firefox patch workflow

1. Find the upstream path with Searchfox.
2. Edit \`.geckokit/firefox/<path>\`.
3. Run \`npm run export -- <path>\`.
4. Run \`npm run import\`.
5. Run \`npm run build:fast\`, or \`npm run build\` for native/build-system changes.

Firefox privileged globals such as \`Services\`, \`ChromeUtils\`, and \`Components\` are available to browser modules. Keep feature modules small and validate changes with \`npm run verify\`.
`,
        'CLAUDE.md': '@/AGENTS.md\n',
        'geckokit.json': json({ firefoxSource: '.geckokit/firefox' }),
        'browser.config.json': json(selected ?? {
            browser: {
                name: displayName,
                binaryName: name,
                vendor: displayName,
                applicationId: `dev.geckokit.${name}`,
            },
            firefox: { channel: 'release', version: 'latest' },
            branding: { enabled: false, directory: 'branding' },
            patches: { directory: 'patches' },
            overlays: { directory: 'src/firefox' },
            features: { preset: 'default', selected: [] },
            build: { mode: 'artifact' },
            language,
        }),
        'package.json': json({
            name,
            private: true,
            scripts: {
                setup: 'geckokit setup',
                build: 'geckokit build',
                'build:fast': 'geckokit build faster',
                dev: 'geckokit dev',
                start: 'geckokit run',
                verify: 'geckokit verify',
                export: 'geckokit export',
                import: 'geckokit import',
                package: 'geckokit package',
            },
            devDependencies: { geckokit: VERSION },
        }),
        '.gitignore': `.geckokit/\n.geckokit-profile/\n`,
        'docs/customizing.md': customizationGuide(displayName),
        'patches/README.md': `# Firefox patches

Use patches for small edits to upstream Firefox files. GeckoKit validates and applies them in filename order before each build.

\`\`\`bash
# Find the path with Searchfox, then edit its local working copy.
$EDITOR .geckokit/firefox/<path>

# Export, reapply, and rebuild.
npm run export -- <path>
npm run import
npm run build:fast
\`\`\`

Use \`src/firefox/<path>\` instead when your project should own the complete file. Use a full build for native code, WebIDL, configure, or build-system changes.
`,
        'src/firefox/README.md': `# Firefox source files

This directory mirrors the Firefox source root. For example, \`src/firefox/browser/components/example.txt\` replaces \`.geckokit/firefox/browser/components/example.txt\` before patches apply.

Use this for complete file ownership. For a small edit to a large upstream file, edit the working copy under \`.geckokit/firefox/\`, then run \`npm run export -- <path>\`.
`,
        'src/browser/README.md': `# Browser app code

- \`${startupFile}\`: privileged code run after the first browser window is ready
- \`features/\`: one module or folder per product feature
- \`views/\`: packaged HTML, CSS, and JavaScript
- \`assets/\`: packaged icons and images
- \`styles/browser.css\`: tabs, toolbars, menus, and other browser chrome
- \`styles/content.css\`: web content and built-in pages

Edit or remove the marked onboarding block in \`${startupFile}\`. Run \`npm run dev\` after the first build. CSS is copied into the development profile and restarts without a rebuild.
`,
        'src/browser/features/README.md': `# Features

Keep one feature per module or folder. Export a small setup function, then import and call it from \`${startupFile}\`.
`,
        'src/browser/views/README.md': `# Views

Put packaged HTML, CSS, and JavaScript here. Open files through \`resource://geckokit/views/\` from privileged feature modules.
`,
        'src/browser/assets/README.md': `# Assets

Put feature icons and images here. Reference them through \`resource://geckokit/assets/\`.
`,
        [`src/browser/${startupFile}`]: `export async function startup(window${language === 'typescript' ? ': any' : ''}) {
  // Remove this onboarding block to disable the welcome UI.
  const seenPref = "geckokit.onboarding.seen";
  if (Services.prefs.getBoolPref(seenPref, false)) {
    return;
  }

  await window.gNotificationBox.appendNotification(
    "geckokit-onboarding",
    {
      label: ${JSON.stringify(`Welcome to ${displayName}. Start with README.md. Edit this UI in src/browser/${startupFile}.`)},
      image: "chrome://branding/content/icon64.png",
      priority: window.gNotificationBox.PRIORITY_INFO_MEDIUM,
    },
    [],
  );
  Services.prefs.setBoolPref(seenPref, true);
}
`,
        ...(language === 'typescript' ? {
            'src/browser/geckokit.d.ts': 'declare const Services: any;\ndeclare const ChromeUtils: any;\ndeclare const Components: any;\ndeclare const Cc: any;\ndeclare const Ci: any;\ndeclare const Cu: any;\n',
        } : {}),
        'src/browser/styles/browser.css': `:root {
  --geckokit-accent: #7c3aed;
}

#PersonalToolbar {
  display: none;
}

#navigator-toolbox {
  border-bottom: 2px solid var(--geckokit-accent) !important;
}

#TabsToolbar,
#nav-bar {
  box-shadow: none !important;
}
`,
        'src/browser/styles/content.css': `@-moz-document url("about:newtab"), url("about:home") {
  :root {
    --newtab-primary-action-background: #7c3aed !important;
  }
}
`,
        'prefs/browser.js': `pref("toolkit.legacyUserProfileCustomizations.stylesheets", true);
`,
        'branding/configure.sh': `MOZ_APP_DISPLAYNAME=${shellQuote(displayName)}
`,
        'branding/README.md': `# Branding

Enable branding in \`browser.config.json\`, then customize:

- \`configure.sh\`: display name
- \`locales/en-US/brand.ftl\`: product and vendor strings
- \`content/\`: icons and visual assets

GeckoKit starts from Firefox's version-matched unofficial branding tree. Replace every inherited asset before distribution.
`,
        'branding/locales/en-US/brand.ftl': `-brand-shorter-name = ${displayName}
-brand-short-name = ${displayName}
-brand-shortcut-name = ${displayName}
-brand-full-name = ${displayName}
-brand-product-name = ${displayName}
-vendor-short-name = ${vendor}
trademarkInfo = { " " }
`,
        'branding/content/icon.svg': `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 64 64">
  <rect width="64" height="64" rx="14" fill="#7c3aed"/>
  <text x="32" y="42" text-anchor="middle" font-family="sans-serif" font-size="32" fill="white">${displayName[0]}</text>
</svg>
`,
    };
}
